crates/flux-refineck/src/checker.rs

```diff
@@ -1039,6 +1067,142 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
         }
     }
 
+    fn closure_template(
+        &mut self,
+        infcx: &mut InferCtxt<'_, 'genv, 'tcx>,
+        env: &mut TypeEnv,
+        stmt_span: Span,
+        args: &flux_rustc_bridge::ty::GenericArgs,
+        operands: &[Operand],
+    ) -> Result<(Vec<Ty>, PolyFnSig)> {
+        let upvar_tys = self
+            .check_operands(infcx, env, stmt_span, operands)
+            .with_span(stmt_span)?
+            .into_iter()
+            .map(|ty| {
+                if let TyKind::Ptr(PtrKind::Mut(re), path) = ty.kind() {
+                    env.ptr_to_ref(
+                        &mut infcx.at(stmt_span),
+                        ConstrReason::Other,
+                        *re,
+                        path,
+                        PtrToRefBound::Infer,
+                    )
+                } else {
+                    Ok(ty.clone())
+                }
+            })
+            .try_collect_vec()
+            .with_span(stmt_span)?;
+
+        let closure_args = args.as_closure();
+        let ty = closure_args.sig_as_fn_ptr_ty();
+
+        if let flux_rustc_bridge::ty::TyKind::FnPtr(poly_sig) = ty.kind() {
+            let poly_sig = poly_sig.unpack_closure_sig();
+            let poly_sig = self.refine_with_holes(&poly_sig).with_span(stmt_span)?;
+            let poly_sig = poly_sig
+                .replace_holes(|binders, kind| infcx.fresh_infer_var_for_hole(binders, kind));
+            Ok((upvar_tys, poly_sig))
+        } else {
+            bug!("check_rvalue: closure: expected fn_ptr ty, found {ty:?} in {args:?}");
+        }
+    }
+
+    fn to_closure_sig(
```

Comment:
Why did you move this function here? It doesn't seem to require anything other than the `tcx`

Reply:
unclear - have moved back to original spot!